### DISCUSSION TITLE
Problems with dielectrics in One Weekend

### DISCUSSION THREAD

**Discussion by a42027756** [RayTracing/raytracing.github.io#997] [Q&A] [answered]
I was doing fine following the instructions and codes of the book 1, but one problem occurred when I began the dielectric chapter.
I followed the code in the book, but the result was like:
![dieletrics](https://user-images.githubusercontent.com/67039529/157798033-fcdacf8b-6845-4d14-b002-0e79546fef30.png)
The dielectric material is just black, and others seem to work just fine.
I tried to add both reflections and refractions to the dielectrics, and it looks like this:
![dieletrics_broken100](https://user-images.githubusercontent.com/67039529/157798275-3be220a8-6d91-47f2-9597-19bdac4b2bc2.png)
It seems that the reflection part is doing just fine, and something's wrong with the refraction.
Does anyone have a clue? 
Here's my code if anyone tries to look it up:
https://github.com/a42027756/RayTraing-in-one-weekend

PS: Sorry for my bad English BTW

**Comment by ghadeeras** [comment, marked as answer]
Hello!

My guess is that you might have a problem in the way you calculate the normal when the ray hits the sphere surface from the inside (that is, after it had refracted for the first time and scattered inside the sphere)

I see in your sphere class, you have two main cases to examine the two intersection points between the ray and the sphere. In one of them, you are not using the normal's setter method, which I believe adjusts the direction of the normal depending on whether the ray hits the surface from the outside or the inside.

Long story short, I think your rays are bouncing infinitely inside the dielectric sphere after they have refracted the first time. If my guess above is not correct, you could debug your code by sending just a single ray towards the sphere and examining things under a debugger.

Hope this helps!

**Reply by a42027756** [reply]
Hey !! Just letting you know that I examined the hit() function in sphere class and modified it, now refraction is working perfectly! 
Thank you so much for enlightening me!!

I guess that if anyone encounters the problem that the material turns black, it is very likely that it is because rays are bouncing infinitely until it reaches the maximum depth, and are set to color(0,0,0), resulting in black color, just like mine. So hopefully others who are suffering from this problem can get some inspirations from this case.

**Reply by NoodleLadJr** [reply]
Had the same problem, my function to set the normal was also not correct, making the material black. Thanks for the answer, it seems pretty hard to debug these types of issues!

**Reply by lyzirving** [reply]
> Hey !! Just letting you know that I examined the hit() function in sphere class and modified it, now refraction is working perfectly! Thank you so much for enlightening me!!
> 
> I guess that if anyone encounters the problem that the material turns black, it is very likely that it is because rays are bouncing infinitely until it reaches the maximum depth, and are set to color(0,0,0), resulting in black color, just like mine. So hopefully others who are suffering from this problem can get some inspirations from this case.

hello,
can you show the exact modification?

**Reply by quadpixels** [reply]
I ran into this as well, the issue was an automatic typecast from float to bool:

In this code:
`bool front_face = glm::dot(r.direction, outward_normal);`

`front_face` will be true if `glm::dot` returns a nonzero value, including negative values.

The correct way is: `bool front_face = glm::dot(r.direction, outward_normal) < 0;`. Notice the `< 0` in the end.

**Reply by 9mm** [reply]
In my case, my `hit` function in `Impl Hittable for Sphere` was shadowing a variable instead of mutating it. The correct code is:

```rust
        let mut root = (h - sqrtd) / a;
        if !t.surrounds(root) {
            root = (h + sqrtd) / a;
            if !t.surrounds(root) {
                return false;
            }
        }
```

**Reply by buxtonpaul** [reply]
@9mm  You absolute star. I had exactly the same issue and was beginning to start pulling out my hair. :-)

**Reply by 9mm** [reply]
@buxtonpaul 🤣 you got it, I know them feels
